[Substrate.io Migration] Tutorials -> Build a Blockchain -> Add Trusted Nodes (#29)
polkadot-developers/polkadot-docs: tutorials/polkadot-sdk/build-a-blockchain/add-trusted-nodes.md

This process of generating your own keys serves several important purposes:

- It enhances the security of your network by ensuring that each node has its own unique cryptographic identity
- It simulates a more realistic blockchain environment where participants don't share key information
- It helps you understand the process of key generation, which is a fundamental skill in blockchain operations

To understand more about the different signing algorithms used in this tutorial (sr25519 and ed25519), check [Keypairs and Signing](https://wiki.polkadot.network/docs/learn-cryptography#keypairs-and-signing){target=\_blank}. To learn more about the different types of keys used, refer to the [Keys](https://wiki.polkadot.network/docs/learn-cryptography#keys){target=\_blank} section in the Polkadot Wiki.

### Key Generation Options

There are several ways you can generate keys. The available methods are:

> - `solochain-template-node` [key](TODO:update-path) subcommand - the most straightforward method for developers working directly with the node is to use the integrated key generation feature. Using the `key` subcommand, you can generate keys directly from your node's command line interface. This method ensures compatibility with your chain and is convenient for quick setup and testing

Reviewer: This seems like it should be `solochain-template-node key`

### Generate Local Keys with the Node Template

Best practices for key generation:

- Whenever possible, use an air-gapped computer, meaning never connected to the internet, when generating keys for a production blockchain
- If an air-gapped computer is not an option, disconnect from the internet before generating keys for any public or private blockchain not under your control

For this tutorial, however, you'll use the `solochain-template-node` command-line options to generate random keys locally while remaining connected to the internet. This method is suitable for learning and testing purposes.

Follow these steps to generate your keys:

1. Navigate to the root directory where you compiled the node template

2. Generate a random secret phrase and Sr25519 keys. Enter a password when prompted:

    ```bash
    ./target/release/solochain-template-node key generate \
    --scheme Sr25519 \
    --password-interactive
    ```

    The command will output information about the generated keys similar to the following:

    --8<-- 'code/tutorials/polkadot-sdk/build-a-blockchain/add-trusted-nodes/key-sr25519-1.html'

>     Note the Sr25519 public key for the account (SS58 format). This key will be used for producing blocks with `aura`. In this example, the Sr25519 public key for the account is `5HMhkSHpD4XcibjbU9ZiGemLpnsTUzLsG5JhQJQEcxp3KJaW`.

Reviewer: is aura referring to code here, otherwise I feel like we should spell it like Aura

- Sr25519 (for Aura) - `5Df9bvnbqKNR8S1W2Uj5XSpJCKUomyymwCGf6WHKyoo3GDev`
- Ed25519 (for Grandpa)  `5DJRQQWEaJart5yQnA6gnKLYKHLdpX6V4vHgzAYfNPT2NNuW`

## Create a Custom Chain Specification

After generating key pairs for your blockchain, the next step is creating a custom chain specification. You will share this specification with trusted validators participating in your network.

To enable others to participate in your blockchain, ensure that each participant generates their own key pair. Once you collect the keys from all network participants, you can create a custom chain specification to replace the default local one.

In this tutorial, you'll modify the local chain specification to create a custom version for a two-node network. The same process can be used to add more nodes if you have the necessary keys.

### Steps to Create a Custom Chain Specification

1. Open a terminal and navigate to the root directory of your compiled node template

2. Export the local chain specification:

    ```bash
    ./target/release/solochain-template-node build-spec \
    --disable-default-bootnode \
    --chain local > customSpec.json
    ```

3. Preview the `customSpec.json` file:

>     - Preview first fields:

Reviewer: Is there a command that should be run here? It should be in triple backticks so users can copy pasta

    --base-path /tmp/node01 \
    --chain customSpecRaw.json \
    --scheme Sr25519 \
    --suri "INSERT_SECRET_PHRASE" \
    --password-interactive \
    --key-type aura
    ```

    !!!note
        Replace `INSERT_SECRET_PHRASE` with the secret phrase or seed you generated earlier. When prompted, enter the password you used to generate the keys.

3. Insert the `grandpa` secret key:
    ```bash
    ./target/release/solochain-template-node key insert \
    --base-path /tmp/node01 \
    --chain customSpecRaw.json \
    --scheme Ed25519 \
    --suri "INSERT_SECRET_PHRASE" \
    --password-interactive \
    --key-type gran
    ```

    !!!note
        Use the same secret phrase or seed and password as in step 2.

4. Verify that your keys are in the keystore by running the following command:

    ```bash
    ls /tmp/node01/chains/local_testnet/keystore
    ```

    You should see output similar to:

    <div id='termynal' data-termynal>
        <span data-ty>61757261ea23fa399c6bd91af3d7ea2d0ad46c48aff818b285342d9aaf15b3172270e914</span>
        <span data-ty>6772616ec9c2cd111f98f2bf78bab6787449fc007dd7f2a5d02f099919f7fb50ade97dd6</span>

>     </div>

Reviewer: I've noticed that the commands haven't been added to these terminal elements, and I've kinda been like ehhh okayyyy, but here in particular looks weird without having the command there 🤷‍♀️

## Start the First Node

Before starting the first node, it's crucial to generate a network key. This key ensures that the node's identity remains consistent, allowing other nodes to connect to it reliably as a bootnode for synchronization.

To generate a network key, run the following command:

```bash
./target/release/solochain-template-node key \
generate-node-key --base-path /tmp/node01
```

!!!note
    This command generates a network key and stores it in the same base path used for storing the `aura` and `grandpa` keys.

After generating the network key, start the first node using your custom chain specification with the following command:

```bash
./target/release/solochain-template-node \
--base-path /tmp/node01 \
--chain ./customSpecRaw.json \
--port 30333 \
--rpc-port 9945 \
--validator \
--name MyNode01 \
--password-interactive
```

Upon execution, you should see output similar to the following:

--8<-- 'code/tutorials/polkadot-sdk/build-a-blockchain/add-trusted-nodes/node-output.html'

After starting the first node, you'll notice:

- The node is running with the custom chain specification ("My Custom Testnet")

> - The local node identity is displayed (12D3KooWSbaPxmb2tWLgkQVoJdxzpBPTd9dQPmKiJfsvtP753Rg1 in this example). This identity is crucial for other nodes to connect to this one

Reviewer: ```suggestion
- The local node identity is displayed (`12D3KooWSbaPxmb2tWLgkQVoJdxzpBPTd9dQPmKiJfsvtP753Rg1` in this example). This identity is crucial for other nodes to connect to this one
```

This process of generating your own keys serves several important purposes:

- It enhances the security of your network by ensuring that each node has its own unique cryptographic identity
- It simulates a more realistic blockchain environment where participants don't share key information
- It helps you understand the process of key generation, which is a fundamental skill in blockchain operations

To understand more about the different signing algorithms used in this tutorial (sr25519 and ed25519), check [Keypairs and Signing](https://wiki.polkadot.network/docs/learn-cryptography#keypairs-and-signing){target=\_blank}. To learn more about the different types of keys used, refer to the [Keys](https://wiki.polkadot.network/docs/learn-cryptography#keys){target=\_blank} section in the Polkadot Wiki.

### Key Generation Options

There are several ways you can generate keys. The available methods are:

> - solochain-template-node [key](TODO:update-path) subcommand - the most straightforward method for developers working directly with the node is to use the integrated key generation feature. You can generate keys directly from your node's command line interface using the `key` subcommand. This method ensures compatibility with your chain and is convenient for quick setup and testing

Reviewer 2: ```suggestion
- `solochain-template-node` [key](TODO:update-path) subcommand - the most straightforward method for developers working directly with the node is to use the integrated key generation feature. Using the `key` subcommand, you can generate keys directly from your node's command line interface. This method ensures compatibility with your chain and is convenient for quick setup and testing
```

- [subkey](https://github.com/paritytech/polkadot-sdk/tree/master/substrate/bin/utils/subkey){target=\_blank} - it is a powerful standalone utility specifically designed for Polkadot SDK-based chains. It offers advanced options for key generation, including support for different key types such as `ed25519` and `sr25519`. This tool allows fine-grained control over the key generation process
- Third-party key generation utilities - various tools developed by the community

### Generate Local Keys with the Node Template

Best practices for key generation:

> - Ideally, use an air-gapped computer (never connected to the internet) when generating keys for a production blockchain

Reviewer 2: ```suggestion
- Whenever possible, use an air-gapped computer, meaning never connected to the internet, when generating keys for a production blockchain
```

> - At minimum, disconnect from the internet before generating keys for any public or private blockchain not under your control

Reviewer 2: ```suggestion
- If an air-gapped computer is not an option, disconnect from the internet before generating keys for any public or private blockchain not under your control
```

    --8<-- 'code/tutorials/polkadot-sdk/build-a-blockchain/add-trusted-nodes/key-sr25519-1.html'

    Note the Sr25519 public key for the account (SS58 format). This key will be used for producing blocks with `aura`. In this example, the Sr25519 public key for the account is `5HMhkSHpD4XcibjbU9ZiGemLpnsTUzLsG5JhQJQEcxp3KJaW`.

3. Use the generated secret phrase to derive keys using the Ed25519 signature scheme. Enter the same password you used in the previous step:

    ```bash
    ./target/release/solochain-template-node key inspect \
    --scheme Ed25519 \
    --password-interactive \
    "INSERT_SECRET_PHRASE"
    ```

    !!! note
        Replace `INSERT_SECRET_PHRASE` with the secret phrase generated in step 2.

    The command will output information about the generated keys similar to the following:

    --8<-- 'code/tutorials/polkadot-sdk/build-a-blockchain/add-trusted-nodes/key-ed25519-1.html'

    The Ed25519 key you've generated is crucial for block finalization using the `grandpa` consensus algorithm. The Ed25519 public key for the account is `5GdFMFbXy24uz8mFZroFUgdBkY2pq6igBNGAq9tsBfEZRSzP`.

### Generate a Second Set of Keys

> In this tutorial, the private network will consist of two nodes, meaning you'll need two distinct sets of keys. You have several options for generating these second set of keys:

Reviewer 2: ```suggestion
In this tutorial, the private network will consist of two nodes, meaning you'll need two distinct sets of keys. You have several options for generating this second set of keys:
```

- Use the keys from one of the predefined accounts
- Follow the steps from the previous section, but use a different identity on your local machine to create a new key pair
- Derive a child key pair to simulate a second identity on your local machine

> For the purpose of this tutorial, the second set of keys will be:

Reviewer 2: ```suggestion
For this tutorial, the second set of keys will be:
```

- Sr25519 (for Aura) - `5Df9bvnbqKNR8S1W2Uj5XSpJCKUomyymwCGf6WHKyoo3GDev`
- Ed25519 (for Grandpa)  `5DJRQQWEaJart5yQnA6gnKLYKHLdpX6V4vHgzAYfNPT2NNuW`

## Create a Custom Chain Specification

> After generating key pairs for your blockchain, the next step is creating a custom chain specification. This specification will be shared with trusted validators participating in your network.

Reviewer 2: ```suggestion
After generating key pairs for your blockchain, the next step is creating a custom chain specification. You will share this specification with trusted validators participating in your network.
```

To enable others to participate in your blockchain, ensure that each participant generates their own key pair. Once you collect the keys from all network participants, you can create a custom chain specification to replace the default local one.

In this tutorial, you'll modify the local chain specification to create a custom version for a two-node network. The same process can be used to add more nodes if you have the necessary keys.

### Steps to Create a Custom Chain Specification

1. Open a terminal and navigate to the root directory of your compiled node template

> 2. Export the local chain specification

Reviewer 2: ```suggestion
2. Export the local chain specification:
```

    ```bash
    ./target/release/solochain-template-node build-spec \
    --disable-default-bootnode \
    --chain local > customSpec.json
    ```

3. Preview the `customSpec.json` file:

>     - Preview first fields

Reviewer 2: ```suggestion
    - Preview first fields:
```

    --8<-- 'code/tutorials/polkadot-sdk/build-a-blockchain/add-trusted-nodes/chainspec-head.html'

>     - Preview last fields

Reviewer 2: ```suggestion
    - Preview last fields:
```

    --8<-- 'code/tutorials/polkadot-sdk/build-a-blockchain/add-trusted-nodes/chainspec-tail.html'

>         This command will display fields that include configuration details for pallets such as sudo and balances, as well as the validator settings for the Aura and Grandpa keys.

Reviewer 2: ```suggestion
        This command will display fields that include configuration details for pallets, such as sudo and balances, as well as the validator settings for the Aura and Grandpa keys.
```

4. Edit `customSpec.json`:

    1. Update the `name` field:
    ```json
    "name": "My Custom Testnet",
    ```

>     2. Modify the `aura` field to specify the nodes with the authority to create blocks:

Reviewer 2: ```suggestion
    2. Add Sr25519 addresses for each validator in the authorities array to the `aura` field to specify the nodes with authority to create blocks:
```

>         -  Add Sr25519 addresses for each validator in the authorities array

Reviewer 2: ```suggestion
```

        ```json
        "aura": {
          "authorities": [
            "5HMhkSHpD4XcibjbU9ZiGemLpnsTUzLsG5JhQJQEcxp3KJaW",
            "5Df9bvnbqKNR8S1W2Uj5XSpJCKUomyymwCGf6WHKyoo3GDev"
          ]
        },
        ```

>     3. Update the `grandpa` field to specify the nodes with the authority to finalize blocks:

Reviewer 2: ```suggestion
    3. Add Ed25519 addresses for each validator in the authorities array to the `grandpa` field to specify the nodes with the authority to finalize blocks. Include a voting weight (typically `1`) for each validator to define their voting power:
```

>         - Add Ed25519 addresses for each validator in the authorities array 

Reviewer 2: ```suggestion
```

>         - Include a voting weight (typically 1) for each validator to define their voting power

Reviewer 2: ```suggestion
```

    --base-path /tmp/node01 \
    --chain customSpecRaw.json \
    --scheme Sr25519 \
    --suri "INSERT_SECRET_PHRASE" \
    --password-interactive \
    --key-type aura
    ```

    !!!note
        Replace `INSERT_SECRET_PHRASE` with the secret phrase or seed you generated earlier. When prompted, enter the password you used to generate the keys.

3. Insert the `grandpa` secret key:
    ```bash
    ./target/release/solochain-template-node key insert \
    --base-path /tmp/node01 \
    --chain customSpecRaw.json \
    --scheme Ed25519 \
    --suri "INSERT_SECRET_PHRASE" \
    --password-interactive \
    --key-type gran
    ```

    !!!note
        Use the same secret phrase or seed and password as in step 2.

4. Verify that your keys are in the keystore by running the following command:

    ```bash
    ls /tmp/node01/chains/local_testnet/keystore
    ```

    You should see output similar to:

    <div id='termynal' data-termynal>
        <span data-ty>61757261ea23fa399c6bd91af3d7ea2d0ad46c48aff818b285342d9aaf15b3172270e914</span>
        <span data-ty>6772616ec9c2cd111f98f2bf78bab6787449fc007dd7f2a5d02f099919f7fb50ade97dd6</span>
    </div>

## Start the First Node

> Before starting the first node, it's crucial to generate a network key. This key ensures that the node's identity remains consistent, allowing other nodes to reliably connect to it as a bootnode for synchronization.

Reviewer 2: ```suggestion
Before starting the first node, it's crucial to generate a network key. This key ensures that the node's identity remains consistent, allowing other nodes to connect to it reliably as a bootnode for synchronization.
```

--port 30333 \
--rpc-port 9945 \
--validator \
--name MyNode01 \
--password-interactive
```

Upon execution, you should see output similar to the following:

--8<-- 'code/tutorials/polkadot-sdk/build-a-blockchain/add-trusted-nodes/node-output.html'

After starting the first node, you'll notice:

- The node is running with the custom chain specification ("My Custom Testnet")
- The local node identity is displayed (12D3KooWSbaPxmb2tWLgkQVoJdxzpBPTd9dQPmKiJfsvtP753Rg1 in this example). This identity is crucial for other nodes to connect to this one
- The node is currently idle with 0 peers, as it's the only node in the network at this point
- No blocks are being produced. Block production will commence once another node joins the network

## Add More Nodes

Block finalization requires at least two-thirds of the validators. In this example network configured with two validators, block finalization can only start after the second node has been added.

Before starting additional nodes, ensure you've properly configured their keys as described in the [Add Keys to the Keystore](#add-keys-to-the-keystore) section. For this node, the keys should be stored under the `/tmp/node02` base path.

> To add a second validator to the private network run the following command:

Reviewer 2: ```suggestion
To add a second validator to the private network, run the following command:
```

--unsafe-force-node-key-generation \
--password-interactive
```

Key points about this command:

- It uses a different `base-path` and name to identify this as the second validator
- The `--chain` option specifies the same chain specification file used for the first node
- The `--bootnodes` option is crucial. It should contain the local node identifier from the first node in the network
- The `--unsafe-force-node-key-generation` parameter forces the generation of a new node key if one doesn't exist. For non-bootnode validators (like this second node and any subsequent nodes), it's less critical if the key changes because they won't be used as bootnodes. However, for consistency and best practices, it's recommended to generate and maintain a stable node key for all validators once the network is set up

After both nodes have added their keys to their respective keystores (under `/tmp/node01` and `/tmp/node02`) and been run, you should see:

- The same genesis block and state root hashes on both nodes
- Each node showing one peer
- Block proposals being produced 
- After a few seconds, new blocks being finalized on both nodes
  

> If every step was correctly performed, you should see logs similar to the following on both nodes:

Reviewer 2: ```suggestion
If successful, you should see logs similar to the following on both nodes:
```

- If an air-gapped computer is not an option, disconnect from the internet before generating keys for any public or private blockchain not under your control

For this tutorial, however, you'll use the `solochain-template-node` command-line options to generate random keys locally while remaining connected to the internet. This method is suitable for learning and testing purposes.

Follow these steps to generate your keys:

1. Navigate to the root directory where you compiled the node template

2. Generate a random secret phrase and Sr25519 keys. Enter a password when prompted:

    ```bash
    ./target/release/solochain-template-node key generate \
    --scheme Sr25519 \
    --password-interactive
    ```

    The command will output information about the generated keys similar to the following:

    --8<-- 'code/tutorials/polkadot-sdk/build-a-blockchain/add-trusted-nodes/key-sr25519-1.html'

    Note the Sr25519 public key for the account (SS58 format). This key will be used for producing blocks with Aura. In this example, the Sr25519 public key for the account is `5HMhkSHpD4XcibjbU9ZiGemLpnsTUzLsG5JhQJQEcxp3KJaW`.

> 3. Use the generated secret phrase to derive keys using the Ed25519 signature scheme. Enter the same password you used in the previous step:
> 
>     ```bash
>     ./target/release/solochain-template-node key inspect \
>     --scheme Ed25519 \
>     --password-interactive \
>     "INSERT_SECRET_PHRASE"
>     ```

Reviewer: ```suggestion
3. Use the generated secret phrase to derive keys using the Ed25519 signature scheme. 

    ```bash
    ./target/release/solochain-template-node key inspect \
    --scheme Ed25519 \
    --password-interactive \
    ```

    When prompted for a **Key password**, enter the same password you used in the previous step
```

Reviewer: I tested it out, the above command doesn't work when you provide the key along with the command. You have to wait til it prompts you

- Derive a child key pair to simulate a second identity on your local machine

For this tutorial, the second set of keys will be:

- Sr25519 (for Aura) - `5Df9bvnbqKNR8S1W2Uj5XSpJCKUomyymwCGf6WHKyoo3GDev`
- Ed25519 (for Grandpa)  `5DJRQQWEaJart5yQnA6gnKLYKHLdpX6V4vHgzAYfNPT2NNuW`

## Create a Custom Chain Specification

After generating key pairs for your blockchain, the next step is creating a custom chain specification. You will share this specification with trusted validators participating in your network.

To enable others to participate in your blockchain, ensure that each participant generates their own key pair. Once you collect the keys from all network participants, you can create a custom chain specification to replace the default local one.

In this tutorial, you'll modify the local chain specification to create a custom version for a two-node network. The same process can be used to add more nodes if you have the necessary keys.

### Steps to Create a Custom Chain Specification

1. Open a terminal and navigate to the root directory of your compiled node template

2. Export the local chain specification:

>     ```bash

Reviewer 3: Note that, while this should still work, we should encourage chain-spec-builder later.

Author: I think we can merge it now and update it later

---
title: Add Trusted Nodes

> description: TODO

Reviewer 4: missing description

This process of generating your own keys serves several important purposes:

- It enhances the security of your network by ensuring that each node has its own unique cryptographic identity
- It simulates a more realistic blockchain environment where participants don't share key information
- It helps you understand the process of key generation, which is a fundamental skill in blockchain operations

### Key Generation Options

### Generate Local Keys with the Node Template

Best practices for key generation:

- Ideally, use an air-gapped computer (never connected to the internet) when generating keys for a production blockchain
- At minimum, disconnect from the internet before generating keys for any public or private blockchain not under your control

For this tutorial, however, we'll use the `solochain-template-node `command-line options to generate random keys locally while remaining connected to the internet. This method is suitable for learning and testing purposes.

Follow these steps to generate your keys:

1. Navigate to the root directory where you compiled the node template

> 2. Generate a random secret phrase and Sr25519 keys. Enter a password when prompted:

Reviewer 2: Hey, @nhussein11! Do you think it needs more than this from the preceding section?

"To understand more about the different signing algorithms used in this tutorial (sr25519 and ed25519), check [Keypairs and Signing](https://wiki.polkadot.network/docs/learn-cryptography#keypairs-and-signing){target=\_blank}. "

If so, can you tell me who is working on updating this piece?

> 2. Generate a random secret phrase and Sr25519 keys. Enter a password when prompted:

Reviewer 4: @dawnkelly09 mm I think that should be enough to contextualize a just a bit...

Maybe link "Keypairs and Signing" with this reference: https://wiki.polkadot.network/docs/learn-cryptography#keypairs-and-signing 

> If so, can you tell me who is working on updating this piece?

I think @0xLucca since he is the author of this PR. @0xLucca if you want me to take this one just let me know :)

> 2. Generate a random secret phrase and Sr25519 keys. Enter a password when prompted:

Author: Thats already included in the Generate an account and keys section:

`To understand more about the different signing algorithms used in this tutorial (sr25519 and ed25519), check [Keypairs and Signing](https://wiki.polkadot.network/docs/learn-cryptography#keypairs-and-signing). To learn more about the different types of keys used, refer to the [Keys](https://wiki.polkadot.network/docs/learn-cryptography#keys) section in the Polkadot Wiki.`

    ```bash
    ./target/release/solochain-template-node key generate --scheme Sr25519 --password-interactive
    ```

    The command will output information about the generated keys similar to the following:

    ```plain
    Secret phrase:       digital width rely long insect blind usual name oyster easy steak spend
      Network ID:        substrate
      Secret seed:       0xc52405d0b45dd856cbf1371f3b33fbde20cb76bf6ee440d12ea15f7ed17cca0a
      Public key (hex):  0xea23fa399c6bd91af3d7ea2d0ad46c48aff818b285342d9aaf15b3172270e914
      Account ID:        0xea23fa399c6bd91af3d7ea2d0ad46c48aff818b285342d9aaf15b3172270e914
      Public key (SS58): 5HMhkSHpD4XcibjbU9ZiGemLpnsTUzLsG5JhQJQEcxp3KJaW
      SS58 Address:      5HMhkSHpD4XcibjbU9ZiGemLpnsTUzLsG5JhQJQEcxp3KJaW
    ```

>     Note the Sr25519 public key for the account. This key will be used for producing blocks with `aura`. In this example, the Sr25519 public key for the account is `5HMhkSHpD4XcibjbU9ZiGemLpnsTUzLsG5JhQJQEcxp3KJaW`.

Reviewer 4: I'd clarify here:

> Note the Sr25519 public key for the account (SS58 format)

> 3. Use the generated secret phrase to derive keys using the Ed25519 signature scheme. Enter the same password you used in the previous step:

Reviewer 4: I'd clarify or add a reference about what Ed25519 is (or at least why you need to generate firstSr25519 and then Ed25519)

    ```bash
    ./target/release/solochain-template-node key inspect --password-interactive --scheme Ed25519 "INSERT_SECRET_PHRASE"
    ```

    !!! note
        Replace `INSERT_SECRET_PHRASE` with the secret phrase generated in step 2.

    The command will output information about the generated keys similar to the following:

    ```plain
    Secret phrase:       digital width rely long insect blind usual name oyster easy steak spend
      Network ID:        substrate
      Secret seed:       0xc52405d0b45dd856cbf1371f3b33fbde20cb76bf6ee440d12ea15f7ed17cca0a
      Public key (hex):  0xc9c2cd111f98f2bf78bab6787449fc007dd7f2a5d02f099919f7fb50ade97dd6
      Account ID:        0xc9c2cd111f98f2bf78bab6787449fc007dd7f2a5d02f099919f7fb50ade97dd6
      Public key (SS58): 5GdFMFbXy24uz8mFZroFUgdBkY2pq6igBNGAq9tsBfEZRSzP
      SS58 Address:      5GdFMFbXy24uz8mFZroFUgdBkY2pq6igBNGAq9tsBfEZRSzP

>     ```

Reviewer 4: I think this should be a terminal element

- Use the keys from one of the predefined accounts
- Follow the steps from the previous section, but use a different identity on your local machine to create a new key pair
- Derive a child key pair to simulate a second identity on your local machine

For the purpose of this tutorial, the second set of keys will be:

- Sr25519 (for Aura) - `5Df9bvnbqKNR8S1W2Uj5XSpJCKUomyymwCGf6WHKyoo3GDev`
- Ed25519 (for Grandpa)  `5DJRQQWEaJart5yQnA6gnKLYKHLdpX6V4vHgzAYfNPT2NNuW`

## Create a Custom Chain Specification

After generating key pairs for your blockchain, the next step is creating a custom chain specification. This specification will be shared with trusted validators participating in your network.

To enable others to participate in your blockchain, ensure that each participant generates their own key pair. Once you collect the keys from all network participants, you can create a custom chain specification to replace the default local one.

In this tutorial, you'll modify the local chain specification to create a custom version for a two-node network. The same process can be used to add more nodes if you have the necessary keys.

### Add Validators

> When creating your custom chain specification, you'll need to add validators by modifying two key sections:

Reviewer 4: At least for me, it's unclear where I need to update/modify the plain chain spec. Maybe worth adding a reference to a custom chainspec or adding more steps here to provide more info

Author: In the Add Validators section, you don't have to add anything in the chain spec; it is more like an explanation. I agree that it is quite confusing

Reviewer 2: Confirmed with Erin, please include the commands in terminal outputs.

1. Aura Section - controls block production
    - Add Sr25519 addresses for each validator
    - Example: `"aura": { "authorities": ["5CfB...", "5CXG..."] }`
  

2. Grandpa Section - controls block finalization
    - Add Ed25519 addresses for each validator
    - Include a voting weight (typically 1 for equal voting power)
    - Example: `"grandpa": { "authorities": [["5Cuq...", 1], ["5Dpd...", 1]] }`

!!!note
    Always use unique keys for each validator to prevent conflicts in block production.

> ### Steps to Create a Custom Specification

Reviewer 4: ```suggestion
### Steps to Create a Custom Chain Specification
```

1. Open a terminal and navigate to the root directory of your compiled node template

> 2. Export the local chain specification:

Reviewer 4: Can't remember if we already discuss this one, but I'd go with no colons for list items

For example:
```suggestion
2. Export the local chain specification
```

        <span data-ty>            [</span>
        <span data-ty>              "5FA9nQDVg267DEd8m1ZypXLBnvN7SFxYwV7ndqSYGiN9TTpu",</span>
        <span data-ty>              1</span>
        <span data-ty>            ],</span>
        <span data-ty>            [</span>
        <span data-ty>              "5GoNkf6WdbxCFnPdAnYYQyCjAKPJgLNxXwPjwTh6DGg6gN3E",</span>
        <span data-ty>              1</span>
        <span data-ty>            ]</span>
        <span data-ty>          ]</span>
        <span data-ty>        },</span>
        <span data-ty>        "sudo": {</span>
        <span data-ty>          "key": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"</span>
        <span data-ty>        }</span>
        <span data-ty>      }</span>
        <span data-ty>    }</span>
        <span data-ty>  }</span>
        <span data-ty>}%</span>
    </div>

        This command will display fields that include configuration details for pallets such as sudo and balances, as well as the validator settings for the Aura and Grandpa keys.

4. Edit `customSpec.json`:

    1. Update the `name` field:
    ```json
    "name": "My Custom Testnet",
    ```

>     2. Modify the `aura` field to specify the nodes with the authority to create blocks by adding the Sr25519 SS58 address keys:

Reviewer 4: ```suggestion
    2. Modify the `aura` field to specify the nodes with the authority to create blocks by adding the Sr25519 address keys in SS58 format
```

    ```json
    "aura": {
      "authorities": [
        "5HMhkSHpD4XcibjbU9ZiGemLpnsTUzLsG5JhQJQEcxp3KJaW",
        "5Df9bvnbqKNR8S1W2Uj5XSpJCKUomyymwCGf6WHKyoo3GDev"
      ]
    },
    ```

>     3. Update the `grandpa` field to specify the nodes with the authority to finalize blocks by adding the Ed25519 SS58 address keys:

Reviewer 4: ```suggestion
    3. Update the `grandpa` field to specify the nodes with the authority to finalize blocks by adding the Ed25519 address keys in SS58
```

    ```json
    "grandpa": {
      "authorities": [
        [
          "5GdFMFbXy24uz8mFZroFUgdBkY2pq6igBNGAq9tsBfEZRSzP",
          1
        ],
        [
          "5DJRQQWEaJart5yQnA6gnKLYKHLdpX6V4vHgzAYfNPT2NNuW",
          1
        ]
      ]
    },
    ```

5. Save and close `customSpec.json`

### Convert Chain Specification to Raw Format

After creating your custom chain specification, the next crucial step is converting it to a raw format. This process is essential because the raw format includes encoded storage keys that nodes use to reference data in their local storage. By distributing a raw chain specification, you ensure that each node in the network stores data using the same storage keys, which is vital for maintaining data integrity and facilitating network synchronization.

To convert your chain specification to the raw format, follow these steps:

1. Navigate to the root directory where you compiled the node template
   

> 2. Run the following command to convert the `customSpec.json` chain specification to the raw format, saving it as `customSpecRaw.json`:

Reviewer 4: This is a bit weird to read for me, maybe we can go with:

```suggestion
Run the following command to convert the plain "customSpec.json" file to raw format and save it as "customSpecRaw.json":
```

Follow these steps for each node in your network:

1. Open a terminal and navigate to the root directory where you compiled the node template

2. Insert the `aura` secret key:

    ```bash
    ./target/release/solochain-template-node key insert \
    --base-path /tmp/node01 \
    --chain customSpecRaw.json \
    --scheme Sr25519 \
    --suri "INSERT_SECRET_PHRASE" \
    --password-interactive \
    --key-type aura
    ```

    !!!note
        Replace `INSERT_SECRET_PHRASE` with the secret phrase or seed you generated earlier. When prompted, enter the password you used to generate the keys.

3. Insert the `grandpa` secret key:
    ```bash
    ./target/release/solochain-template-node key insert \
    --base-path /tmp/node01 \
    --chain customSpecRaw.json \
    --scheme Ed25519 \
    --suri "INSERT_SECRET_PHRASE" \
    --password-interactive \
    --key-type gran
    ```

    !!!note
        Use the same secret phrase or seed and password as in step 2.

4. Verify that your keys are in the keystore by running the following command:

    ```bash
    ls /tmp/node01/chains/local_testnet/keystore
    ```

    You should see output similar to:

    <div id='termynal' data-termynal>

>         <span data-ty='input'><span class='file-path'></span>ls /tmp/node01/chains/local_testnet/keystore</span>

Reviewer 4: Actually, you provided feedback that it's unnecessary to include the command within the terminal element if it has been shown above

Author: We should define a standard for this, I use to skip the command when it is too long

```

Key points about this command:

- It uses a different `base-path` and name to identify this as the second validator
- The `--chain` option specifies the same chain specification file used for the first node
- The `--bootnodes` option is crucial. It should contain the local node identifier from the first node in the network
- The `--unsafe-force-node-key-generation` parameter forces the generation of a new node key if one doesn't exist. For non-bootnode validators (like this second node and any subsequent nodes), it's less critical if the key changes because they won't be used as bootnodes. However, for consistency and best practices, it's recommended to generate and maintain a stable node key for all validators once the network is set up

After both nodes have added their keys to their respective keystores (under `/tmp/node01` and `/tmp/node02`) and been run, you should see:

- The same genesis block and state root hashes on both nodes
- Each node showing one peer
- Block proposals being produced 
- After a few seconds, new blocks being finalized on both nodes
  
If every step was correctly performed, you should see logs similar to the following on both nodes:

> <div id='termynal' data-termynal>
>     <span data-ty>2024-09-12 15:37:05 💤 Idle (0 peers), best: #0 (0x8af7…53fd), finalized #0 (0x8af7…53fd), ⬇ 0 ⬆ 0</span>
>     <span data-ty>2024-09-12 15:37:08 discovered: 12D3KooWMaL5zqYiMnVikaYCGF65fKekSPqXGgyz92eRcqcnfpey /ip4/192.168.1.2/tcp/30334</span>
>     <span data-ty>2024-09-12 15:37:10 💤 Idle (1 peers), best: #0 (0x8af7…53fd), finalized #0 (0x8af7…53fd), ⬇ 0.6kiB/s ⬆ 0.6kiB/s</span>
>     <span data-ty>2024-09-12 15:37:12 🙌 Starting consensus session on top of parent 0x8af7c72457d437486fe697b4a11ef42b26c8b4448836bdb2220495aea39f53fd (#0)</span>
>     <span data-ty>2024-09-12 15:37:12 🎁 Prepared block for proposing at 1 (6 ms) [hash: 0xb97cb3a4a62f0cb320236469d8e1e13227a15138941f3c9819b6b78f91986262; parent_hash: 0x8af7…53fd; extrinsics (1): [0x1ef4…eecb]</span>
>     <span data-ty>2024-09-12 15:37:12 🔖 Pre-sealed block for proposal at 1. Hash now 0x05115677207265f22c6d428fb00b65a0e139c866c975913431ddefe291124f04, previously 0xb97cb3a4a62f0cb320236469d8e1e13227a15138941f3c9819b6b78f91986262.</span>
>     <span data-ty>2024-09-12 15:37:12 🏆 Imported #1 (0x8af7…53fd → 0x0511…4f04)</span>
>     <span data-ty>2024-09-12 15:37:15 💤 Idle (1 peers), best: #1 (0x0511…4f04), finalized #0 (0x8af7…53fd), ⬇ 0.5kiB/s ⬆ 0.6kiB/s</span>
>     <span data-ty>2024-09-12 15:37:18 🏆 Imported #2 (0x0511…4f04 → 0x17a7…a1fd)</span>
>     <span data-ty>2024-09-12 15:37:20 💤 Idle (1 peers), best: #2 (0x17a7…a1fd), finalized #0 (0x8af7…53fd), ⬇ 0.6kiB/s ⬆ 0.5kiB/s</span>
>     <span data-ty>2024-09-12 15:37:24 🙌 Starting consensus session on top of parent 0x17a77a8799bd58c7b82ca6a1e3322b38e7db574ee6c92fbcbc26bbe5214da1fd (#2)</span>
>     <span data-ty>2024-09-12 15:37:24 🎁 Prepared block for proposing at 3 (1 ms) [hash: 0x74d78266b1ac2514050ced3f34fbf98a28c6a2856f49dbe8b44686440a45f879; parent_hash: 0x17a7…a1fd; extrinsics (1): [0xe35f…8d48]</span>
>     <span data-ty>2024-09-12 15:37:24 🔖 Pre-sealed block for proposal at 3. Hash now 0x12cc1e9492988cfd3ffe4a6eb3186b1abb351a12a97809f7bae4a7319e177dee, previously 0x74d78266b1ac2514050ced3f34fbf98a28c6a2856f49dbe8b44686440a45f879.</span>
>     <span data-ty>2024-09-12 15:37:24 🏆 Imported #3 (0x17a7…a1fd → 0x12cc…7dee)</span>
>     <span data-ty>2024-09-12 15:37:25 💤 Idle (1 peers), best: #3 (0x12cc…7dee), finalized #1 (0x0511…4f04), ⬇ 0.5kiB/s ⬆ 0.6kiB/s</span>
> </div>

Reviewer 4: ```suggestion
<div id='termynal' data-termynal>
    <span data-ty>...</span>
    <span data-ty>2024-09-12 15:37:05 💤 Idle (0 peers), best: #0 (0x8af7…53fd), finalized #0 (0x8af7…53fd), ⬇ 0 ⬆ 0</span>
    <span data-ty>2024-09-12 15:37:08 discovered: 12D3KooWMaL5zqYiMnVikaYCGF65fKekSPqXGgyz92eRcqcnfpey /ip4/192.168.1.2/tcp/30334</span>
    <span data-ty>2024-09-12 15:37:10 💤 Idle (1 peers), best: #0 (0x8af7…53fd), finalized #0 (0x8af7…53fd), ⬇ 0.6kiB/s ⬆ 0.6kiB/s</span>
    <span data-ty>2024-09-12 15:37:12 🙌 Starting consensus session on top of parent 0x8af7c72457d437486fe697b4a11ef42b26c8b4448836bdb2220495aea39f53fd (#0)</span>
    <span data-ty>2024-09-12 15:37:12 🎁 Prepared block for proposing at 1 (6 ms) [hash: 0xb97cb3a4a62f0cb320236469d8e1e13227a15138941f3c9819b6b78f91986262; parent_hash: 0x8af7…53fd; extrinsics (1): [0x1ef4…eecb]</span>
    <span data-ty>2024-09-12 15:37:12 🔖 Pre-sealed block for proposal at 1. Hash now 0x05115677207265f22c6d428fb00b65a0e139c866c975913431ddefe291124f04, previously 0xb97cb3a4a62f0cb320236469d8e1e13227a15138941f3c9819b6b78f91986262.</span>
    <span data-ty>2024-09-12 15:37:12 🏆 Imported #1 (0x8af7…53fd → 0x0511…4f04)</span>
    <span data-ty>2024-09-12 15:37:15 💤 Idle (1 peers), best: #1 (0x0511…4f04), finalized #0 (0x8af7…53fd), ⬇ 0.5kiB/s ⬆ 0.6kiB/s</span>
    <span data-ty>2024-09-12 15:37:18 🏆 Imported #2 (0x0511…4f04 → 0x17a7…a1fd)</span>
    <span data-ty>2024-09-12 15:37:20 💤 Idle (1 peers), best: #2 (0x17a7…a1fd), finalized #0 (0x8af7…53fd), ⬇ 0.6kiB/s ⬆ 0.5kiB/s</span>
    <span data-ty>2024-09-12 15:37:24 🙌 Starting consensus session on top of parent 0x17a77a8799bd58c7b82ca6a1e3322b38e7db574ee6c92fbcbc26bbe5214da1fd (#2)</span>
    <span data-ty>2024-09-12 15:37:24 🎁 Prepared block for proposing at 3 (1 ms) [hash: 0x74d78266b1ac2514050ced3f34fbf98a28c6a2856f49dbe8b44686440a45f879; parent_hash: 0x17a7…a1fd; extrinsics (1): [0xe35f…8d48]</span>
    <span data-ty>2024-09-12 15:37:24 🔖 Pre-sealed block for proposal at 3. Hash now 0x12cc1e9492988cfd3ffe4a6eb3186b1abb351a12a97809f7bae4a7319e177dee, previously 0x74d78266b1ac2514050ced3f34fbf98a28c6a2856f49dbe8b44686440a45f879.</span>
    <span data-ty>2024-09-12 15:37:24 🏆 Imported #3 (0x17a7…a1fd → 0x12cc…7dee)</span>
    <span data-ty>2024-09-12 15:37:25 💤 Idle (1 peers), best: #3 (0x12cc…7dee), finalized #1 (0x0511…4f04), ⬇ 0.5kiB/s ⬆ 0.6kiB/s</span>
     <span data-ty>...</span>
</div>
```
